src/arithmetic/path_funcs/path_funcs.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2018-2019 Redis Labs Ltd. and Contributors
+ *
+ * This file is available under the Redis Labs Source Available License Agreement
+ */
+
+#include "path_funcs.h"
+#include "../func_desc.h"
+#include "../../datatypes/sipath.h"
+#include "../../ast/ast.h"
+#include "../../util/rmalloc.h"
+#include "../../util/arr.h"
+#include "assert.h"
+
+/**
+ * @brief  Appends node into a path object during path building.
+ * @param  path: A pointer to path objects.
+ * @param  n: Node.
+ * @retval None
+ */
+static void _PathBuilder_AppendNode(Path *path, Node *node) {
+	Path_AppendNode(path, *node);
+	uint edge_count = Path_EdgeCount(*path);
+	// If there are edges in the path, verify the node is referenced in the last edge.
+#ifndef NDEBUG
+	if(edge_count > 0) {
+		Edge *e = &path->edges[edge_count - 1];
+		EntityID nodeId = ENTITY_GET_ID(node);
+		assert(nodeId == e->srcNodeID || nodeId == e->destNodeID);
+	}
+#endif
+}
+
+/**
+ * @brief  Reverse the direction of an edge.
+ * @note   A new copy of the edge is created, in order not to invalidate any data in the original edge
+ *         in case it required elsewhere.
+ * @param  *e: Pointer to the original edge.
+ * @retval New edge with inverse direction.
+ */
+static Edge *_Edge_ReverseDirection(Edge *e) {
+	Edge *clone = Edge_Clone(e);
+	Edge_SetSrcNode(clone, e->dest);
+	Edge_SetDestNode(clone, e->src);
+	return clone;
+}
+
+/**
+ * @brief  Appends an edge to path object during path build.
+ * @note   The edge should be added after its source node has been inserted to its right position in the path.
+ *         Edges insertion is done by interliving nodes and edges.
+ * @param  path: Path.
+ * @param  e: Edge.
+ * @param  RTLEdge: Indicates the if the edge is incoming (RTL in query) edge or not.
+ * @retval None
+ */
+static void _PathBuilder_AppendEdge(Path *path, Edge *edge, bool RTLEdge) {
+	assert(Path_NodeCount(*path) > 0);
+	// The edge should connect nodes[edge_count] to nodes[edge_count+1]
+	uint edge_count = Path_EdgeCount(*path);
+	Node n = (*path).nodes[edge_count];
+	EntityID nId = ENTITY_GET_ID(&n);
+	// Validate source node is in the right place.
+	assert(nId == edge->srcNodeID || nId == edge->destNodeID);
```

Comment:
Isn't `nId == edge->destNodeID` indicates something regarding the edge direction?